Major Release

rgomezp opened the pull request:
## Major Release
This PR upgrades this SDK to use the native SDK versions for iOS & Android of 3.0 & 4.0 respectively.

The migration guide can be found [here](https://documentation.onesignal.com/docs/mobile-2020-api-migration-guide).

## API Additions
* `setAppId` function
* Observers:
  * `addPermissionObserver`
  * `addSubscriptionObserver`
  * `addEmailSubscriptionObserver`
* `setNotificationWillShowInForegroundHandler`
* `setNotificationOpenedHandler`
* `disablePush`
* `getDeviceState`
* `deleteTags` (iOS) <- this was missed in the iOS bridge in previous releases

## API Removals
* `configure` (deprecated)
* `init` (no longer needed, this is done automatically)
* `addEventListener` (replaced by observers)
* `getPermissionSubscriptionState`
* `checkPermissions`

## Other Changes
* general refactoring
* clean up of confusing event strings
* modularization
* moved code into new `src` directory
* created `EventManager` helper class to manage events, listeners, and handlers related to those events
* created `NotificationReceivedEvent` class to encapsulate the event & package it with the platform-specific payload as well as add a `complete` function prior to passing into the handler for the `notificationWillShowInForeground` event
* created `helpers` file for various miscellaneous helper functions
* updated example app

## Before merging
* [x] Remove commit intended to facilitate diff-checking `index.js` since it was moved into a different directory
* [ ] Update `deleteTags` reference in OneSignal docs

rgomezp committed:
Index Implementation

* Clean up Events
* Removed `addEventListener`
* Added Observers
* Notification handlers

rgomezp committed:
RNOneSignal java bridge implementation

rgomezp committed:
RNNotificationJob implementation

rgomezp committed:
Removed `RNNotificationJob`

After discussion, we decided we didn't need this class anymore to pursue the following strategy:

For the setNotificationWillShowInForegroundHandler,
* We want to store the callback on the JS side
* We want to store the notification job on the native side
* We can then have a separate function on the JS side that will call a native bridge method to complete the notification job
* We want to store the notification jobs on the native side with UUIDs since multiple notifications can come in

rgomezp committed:
Update iOS EventEmitter

rgomezp committed:
NotificationReceivedEvent Implementation

Encapsulates the notification data as well as the `complete` function which triggers the completion of the notification display. Used in the `notificationWillShowInForegroundHandler`

rgomezp committed:
EventManager Implementation

Manages everything related to events including the event emitter.
- `setupListeners`
- `clearListeners`
- `setEventHandler`
- `generateEventListener`
- `getFinalPayload`

rgomezp committed:
File with event string constants

rgomezp committed:
Package Update

* Updated main (index.js) location in package.json
* Added jsconfig file for VS Code detection

rgomezp committed:
Helpers Functions Implementation

rgomezp committed:
RCTOneSignalEventEmitter implmentation

rgomezp committed:
iOS - RCTOneSignal implementation

rgomezp committed:
iOS: update OneSignal header

rgomezp committed:
Update Example App

carloscuesta committed:
chore(ios): Remove VALID_ARCHS as User-Defined setting on RCTOneSignal

rgomezp committed:
Nits, remove extra whitespace

rgomezp committed:
Delete Tags Implementation

For some reason, we never added the `deleteTags` method to the iOS bridge.
For consistency, I removed the native call in the Android bridge and implemented `deleteTags` via the use of setting empty string with the `sendTags` function.

jkasten2 commented on examples/RNOneSignal/App.js:
```diff
@@ -66,80 +66,81 @@ export default class App extends Component {
-//        OneSignal.removeEventListener('permission', this.onPermissionChange);
-        OneSignal.removeEventListener('emailSubscription', this.onEmailSubscriptionChange);
-        OneSignal.removeEventListener('inAppMessageClicked', this.onInAppMessageClicked);
+        OneSignal.clearHandlers();
```
Wondering in testing if we observed if still need to clean up handlers?

rgomezp replied:
There's two things the event manager tracks: 1) the RN listeners & 2) the JS side handlers stored in a Map object

When I tried logging the contents of the listeners array, there was nothing in there for some reason. I found this very strange, especially given this example:
```js
import { NativeEventEmitter, NativeModules } from 'react-native';
...

  componentDidMount() {
    ...
    const eventEmitter = new NativeEventEmitter(NativeModules.ToastExample);
    this.eventListener = eventEmitter.addListener('EventReminder', (event) => {
       console.log(event.eventProperty) // "someValue"
    });
    ...
  }

  componentWillUnmount() {
    this.eventListener.remove(); //Removes the listener
  }
```
in the [official docs](https://reactnative.dev/docs/native-modules-android#sending-events-to-javascript).

That said, I did notice for the handlers there can be multiple instances of (i.e. the adders), it seems that these _are_ stored somewhere in memory even after the component re-mounts. Note that I don't know that this would impact a production environment, but nevertheless this would lead to confusion in a testing environment where saving the project auto-reloads the app via the metro-bundler. 

I added the `clearHandlers` function for use in the top-level component's `componentWillUnmount`

jkasten2 commented on src/index.js:
```diff
@@ -0,0 +1,424 @@
+'use strict';
```
This new `src/index.js` looks to be the same as `index.js`. Which file is actually being used? Can we delete one?

rgomezp replied:
Correct, we will delete the top-level one before merging. I left it there for ease of diff-checking as moving it into the `src` directory was not picked up by git as a change to the same file

jkasten2 commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -383,13 +396,14 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
     [OneSignal removeTriggerForKey:key];
 }
 
+// to do, check this still works
```
Looks like we still need to test this.

rgomezp replied:
don't recall why I commented this

rgomezp replied:
reason was I switched the `RCT_REMAP_METHOD` to `RCT_EXPORT_METHOD` but it has to stay as is

jkasten2 replied:
I don't see a change here, looks like it has been and still is `RCT_REMAP_METHOD`. 
Since nothing changed with this it seems like we can just remove this comment then.

jkasten2 commented on index.js:
```diff
@@ -216,57 +126,75 @@ export default class OneSignal {
+    static getDeviceState() {
+        // TODO: check if we need to return promise?
+        if (!checkIfInitialized(RNOneSignal)) return Promise.resolve();
+        return RNOneSignal.getDeviceState();
```
There is one difference between IOS and Android we need to match up on.
Android - `areNotificationsEnabled` 
iOS - `hasNotificationPermission`

We talked about `hasNotificationPermission` being the most clear as a shared name for both iOS and android so we need to do the translational here.

jkasten2 replied:
^ Same logic applies to the permission observer

rgomezp replied:
ok. is this what we talked about the naming standards being different for across the native platforms?

rgomezp replied:
> ^ Same logic applies to the permission observer

Can you be more specific? The permission observer returns a different class with different names

jkasten2 replied:
That is it, looks good

jkasten2 commented on index.js:
```diff
@@ -216,57 +126,75 @@ export default class OneSignal {
 
-    static getPermissionSubscriptionState(callback) {
-        if (!checkIfInitialized()) return;
+    static getPermissionSubscriptionState(handler) {
```
Remove, as it was replaced by `getDeviceState`

rgomezp replied:
Done.

jkasten2 commented on index.js:
```diff
@@ -216,57 +126,75 @@ export default class OneSignal {
+        return RNOneSignal.getDeviceState();
+    }
+
+    static checkPermissions(handler) {
```
Remove, I believe the value they need can be retreived from `getDeviceState`

rgomezp replied:
Done.

jkasten2 commented on index.js:
```diff
@@ -288,19 +216,29 @@ export default class OneSignal {
     static deleteTag(key) {
-        if (!checkIfInitialized()) return;
+        if (!checkIfInitialized(RNOneSignal)) return;
+        RNOneSignal.sendTag(key, "");
```
Should keep `RNOneSignal.deleteTag(key);` as it accounts for this implementation detail

rgomezp replied:
what implementation detail?

rgomezp replied:
switched to use native deleteTag method anyway

jkasten2 commented on index.js:
```diff
@@ -288,19 +216,29 @@ export default class OneSignal {
+        tagKeys.forEach((key)=>{
+            finalTags[key] = "";
+        })
+        OneSignal.sendTags(finalTags);
```
Could use `RNOneSignal.deleteTags(tagKeys);` instead.

rgomezp replied:
Done.

jkasten2 commented on index.js:
```diff
@@ -288,19 +216,29 @@ export default class OneSignal {
 
+    /* V I B R A T I O N */
+
     static enableVibrate(enable) {
```
Remove, no longer in native SDK

rgomezp replied:
Done.

jkasten2 commented on index.js:
```diff
@@ -310,7 +248,7 @@ export default class OneSignal {
     }
 
     static enableSound(enable) {
```
Remove, no longer in native SDK.

rgomezp replied:
Done

jkasten2 reviewed (changes requested):
Duplicated index.js, some functions need to be cleaned up yet, some questions on implementation choices, and permission naming between iOS and Android.

I left most comments on `index.js` vs  `src/index.js` but wasn't sure which one we want to keep

rgomezp committed:
Remove `checkPermissions` & `getPermissionSubscriptionState`

Motivation:
These functions are now covered by `getDeviceState` functionality

rgomezp committed:
Removed `enableVibrate` & `enableSound`

Motivation: this was removed in the native SDK

emawby commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -13,6 +13,8 @@
 @implementation RCTOneSignalEventEmitter {
     BOOL hasListeners;
+    NSMutableDictionary* notificationCompletionCache;
+    NSMutableDictionary* receivedNotificationCache;
```
In general instance variables like these should start with an underscore in objc so notificationCompletionCache -> _notificationCompletionCache. (Same applies to hasListeners)

emawby commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -173,6 +154,42 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
+    }];
+}
+
+RCT_EXPORT_METHOD(completeNotificationJob:(NSString*)notificationId) {
```
This method needs to take a boolean for silencing the notification or not. The way the complete method works is that passing a notification the completion handler means it will show, and passing nil to the handler means it will be silenced. We can't just pass nil for the notificationId to this method because then we can't get the completion block. If we pass in a boolean from the javascript we can use that to decide if we provide a notification to the completion block or just pass nil.

Additionally we should add some protections to this method to prevent crashes. If the user sends an notificationId that isn't in our dictionary we will crash when calling the completion block.

rgomezp committed:
iOS: Renamed Completion Method / Added Underscores to Instance Vars

Motivation:
   * Because the completion block needs to be called with _something_ otherwise it will auto-show after 25 seconds, we introduce a second parameter that's a boolean to dictate whether the notification should be displayed or should be silent.
   * We should also protect the completion block from attempting to fire if it is not found in the completion block cache
   * Cache cleanup: after the notification is either fired or silenced, we should clean the notification & corresponding completion block from our cache

rgomezp committed:
Android: rename notif complete method / add display control BOOL & more

Motivation:
   * We moved the event object sent to JS in the `notificationWillShowInForeground` override method a level up since there is no need for the entire notificationReceivedEvent on the JS wrapper side
   * We added a `shouldDisplay` boolean to control display vs silencing of notifs
   * We now check that the cached `receivedEvent` pulled via the notif UUID is non-null and log an error if so

rgomezp committed:
Deduplicate Observers, added `onCatalystInstanceDestroy` Lifecycle Func

Motivation:
* This commit adds logic to de-duplicate observers from being added multiple times. * We also take advantage of the `onCatalystInstanceDestroy` React Native lifecycle function to remove all handlers and observers. This prevents the observers from staying mounted after reloading in a dev environment.

rgomezp committed:
Added error logging for functions with easy-to-mistaken argument types

Motivation:
Some functions names are not obvious what the argument type should be or what it should look like. To facilitate this process as much as possible, we can do a simple type/format check in certain functions and log an error if the argument is incorrect.

emawby commented on ios/RCTOneSignal/RCTOneSignalEventEmitter.m:
```diff
@@ -173,6 +154,54 @@ + (void)sendEventWithName:(NSString *)name withBody:(NSDictionary *)body {
+RCT_EXPORT_METHOD(setNotificationWillShowInForegroundHandler) {
+    [OneSignal setNotificationWillShowInForegroundHandler:^(OSNotification *notif, OSNotificationDisplayResponse completion) {
+        self->_receivedNotificationCache[notif.notificationId] = notif;
+        self->_notificationCompletionCache[notif.notificationId] = completion;
```
Did Xcode require you use self-> here? I don't think we need the self arrow you should just be able to use the instance variable. Other than that it looks good!

rgomezp replied:
It says: `Block implicitly retains 'self'; explicitly mention 'self' to indicate this is intended behavior` @emawby

jkasten2 commented on examples/RNOneSignal/App.js:
```diff
@@ -66,80 +66,81 @@ export default class App extends Component {
+        })
+
+        /* O N E S I G N A L  H A N D L E R S */
+        OneSignal.setNotificationWillShowInForegroundHandler(notifReceived => {
```
Should be `notifReceivedEvent`

rgomezp replied:
Done.

jkasten2 commented on examples/RNOneSignal/App.js:
```diff
@@ -66,80 +66,81 @@ export default class App extends Component {
+        /* O N E S I G N A L  H A N D L E R S */
+        OneSignal.setNotificationWillShowInForegroundHandler(notifReceived => {
+            console.log("OneSignal: notification will show in foreground:", notifReceived);
+            setTimeout(()=>notifReceived.complete(notifReceived), 0);
```
Should be `notifReceived.complete(notifReceivedEvent.getNotification())`

rgomezp replied:
Resolved  w/ c01f866

jkasten2 commented on index.js:
```diff
@@ -1,201 +1,114 @@
-        // Make native request to init notification opened handler
-        if (type === NOTIFICATION_OPENED_EVENT) {
-            RNOneSignal.initNotificationOpenedHandlerParams();
+        if (Platform.OS === 'android') {
```
Remove android check, this method is available for iOS as well.

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -1,201 +1,114 @@
-            } else if (Platform.OS === 'ios') {
-                RNOneSignal.setInAppMessageClickHandler();
-            }
+        if (Platform.OS === 'android') {
```
Remove android check, this method is available for iOS as well.

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -1,201 +1,114 @@
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
+
+        if (Platform.OS === 'android') {
```
Remove android check, this method is available for iOS as well.

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -1,201 +1,114 @@
 
+    /* R E G I S T R A T I O N  E T C */
+
     static registerForPushNotifications() {
```
Remove `registerForPushNotifications` method, it does the same thing as `promptForPushNotificationsWithUserResponse`

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -1,201 +1,114 @@
         }
     }
 
     static requestPermissions(permissions) {
```
Remove `requestPermissions` method, it does the same thing as `promptForPushNotificationsWithUserResponse`

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -212,61 +125,73 @@ export default class OneSignal {
-    static init(appId, iOSSettings) {
-        if (!checkIfInitialized()) return;
-
+    static promptForPushNotificationPermissions(handler) {
```
Remove `promptForPushNotificationPermissions` method, it does the same thing as `promptForPushNotificationsWithUserResponse`

rgomezp replied:
Done

jkasten2 commented on index.js:
```diff
@@ -384,155 +272,146 @@ export default class OneSignal {
-        if (callback === undefined)
-            callback = function(){};
+        if (Platform.OS === 'android') {
+            RNOneSignal.initInAppMessageClickHandlerParams();
```
Can we do everything in the Java wrapper when `setInAppMessageClickHandler` is called so we don't need this android check?

rgomezp replied:
I assume you mean the ObjC wrapper? We could add an empty stub for it

rgomezp replied:
Investigate in the future what `initInAppMessageClickHandlerParams` is used for

jkasten2 commented on ios/RCTOneSignal/RCTOneSignal.m:
```diff
@@ -42,101 +42,19 @@ + (RCTOneSignal *) sharedInstance {
-    
+
     didInitialize = true;
     [OneSignal addSubscriptionObserver:self];
```
In Android we added these observers only when the developer calls them. We should do the same for iOS to keep them consistent.

rgomezp replied:
Done.

jkasten2 commented on ios/RCTOneSignal/RCTOneSignal.m:
```diff
@@ -145,17 +63,47 @@ - (NSDictionary *)jsonObjectWithString:(NSString *)jsonString {
+    [self sendEvent:OSEventString(EmailSubscriptionChanged) withBody:stateChanges.to.toDictionary];
+}
+
+- (void)didBeginObserving {
```
This method looks like it isn't being call anymore. Do we still need it?

rgomezp replied:
I don't thinks so

rgomezp replied:
Removed.

jkasten2 commented on src/events.js:
```diff
@@ -0,0 +1,7 @@
+export const NOTIFICATION_WILL_SHOW = 'OneSignal-notificationWillShowInForeground';
+export const NOTIFICATION_OPENED    = 'OneSignal-remoteNotificationOpened'
+export const IN_APP_MESSAGE_CLICKED = 'OneSignal-inAppMessageClicked';
+export const EMAIL_SUBSCRIPTION_CHANGED = 'OneSignal-emailSubscriptionChanged';
```
nit, group together `_CHANGED` consts

rgomezp replied:
Done.

rgomezp committed:
Code Cleanup, Exposed Observer Methods in ObjC Bridge

rgomezp committed:
Removed Platform Checks in Index.js After Adding Observers to iOS

rgomezp committed:
New `OSNotification` class & `NotificationReceivedEvent` encapsulation

Motivation:
We want to be able to modify the notification in the future via the `notificationWillShowInForegroundHandler`.  To do this, we can create an extra class `OSNotification` to better match with the native API (at least for Java).

See App.js in this commit for an example

rgomezp committed:
Update RCTOneSignalEventEmitter to Correctly use References

Motivation:
Because this code is within a block, we need to be able to create a weak and then strong reference to "self" in order for those changes to occur on the RCTOneSignalEventEmitter object.

rgomezp committed:
iOS: Add Checks So That Observers Aren't Being Added Multiple Times

jkasten2 commented on examples/RNOneSignal/App.js:
```diff
@@ -66,80 +66,86 @@ export default class App extends Component {
+        OneSignal.setNotificationWillShowInForegroundHandler(notifReceivedEvent => {
+            console.log("OneSignal: notification will show in foreground:", notifReceivedEvent);
+            let notif = notifReceivedEvent.getNotification();
+            // future: modify the notification
```
let's omit this from the example for now, since this isn't planned immediately after the major release. There's a good chunk of work on the native SDK side that is needed to support this.

rgomezp replied:
Done

jkasten2 reviewed (commented):
Just one nit and I think this looks good for a beta1 release.

Before that we need to fix the duplicate `index.js`

For the release process we should do the following:
1. Make a branch from this one and call it something like `major_release_4.0.0-beta1-cut`
2. Update native SDKs to latest betas. update `build.gradle` and the `.podspec` files
3. Update Update version in `package.json` to `4.0.0-beta`
4. Ensure the version format works for a beta release process with npmjs
5. Make a new release here: https://github.com/OneSignal/react-native-onesignal/releases/new
6. Publish to npm as a beta

rgomezp commented:
> Just one nit and I think this looks good for a beta1 release.
> 
> Before that we need to fix the duplicate `index.js`
> 
> For the release process we should do the following:
> 
> 1. Make a branch from this one and call it something like `major_release_4.0.0-beta1-cut`
> 2. Update native SDKs to latest betas. update `build.gradle` and the `.podspec` files
> 3. Update Update version in `package.json` to `4.0.0-beta`
> 4. Ensure the version format works for a beta release process with npmjs
> 5. Make a new release here: https://github.com/OneSignal/react-native-onesignal/releases/new
> 6. Publish to npm as a beta

Opened PR: https://github.com/OneSignal/react-native-onesignal/pull/1057

rgomezp committed:
Update `RNOneSignal.java`: remove unused imports, variables, 1 typo nit

* Removed unused imports
* Removed `OSPermissionSubscriptionState` which was removed native-side
* A couple nits

rgomezp committed:
Update Example App

* Commented out `NotificationNotDisplayingExtender` for now

rgomezp committed:
Update package.json versions

rgomezp committed:
Update Native SDK Versions

rgomezp committed:
Match Format Returned to Javascript from iOS Bridge for State Change

Motivation: we want this format to match the one returned by the Android bridge as well

rgomezp committed:
Update Lock Files

rgomezp committed:
Remove duplicate`actionButtons` parameter

rgomezp committed:
Fix Initialization in iOS bridge

* Was missing `initWithLaunchOptions`
* Removed exposed `initialize` method from event emitter bridge
* Updated `initOneSignal` method to accept the launch options passed to it via the `UIApplication+RCTOneSignal.m` file

rgomezp committed:
Update `postNotification` in Java Bridge

Motivation:
   * We want to simplify the `postNotification` function.
   * The function now takes a json object string and a success and failure callback

rgomezp committed:
Update Index.js

Motivation: change the API to be more consistent with Cordova & other wrappers.

We are now only taking 3 arguments: 1) json object string, 2) success + 3) failure callbacks

rgomezp committed:
Update Example App

Motivation: update example app to have a "Post Notification" testing button

rgomezp committed:
Update `postNotification` in iOS Bridge

Motivation:
   * We want to simplify the `postNotification` function.
   * The function now takes a json object string and uses `postNotificationWithJsonString` to send a notification

rgomezp committed:
Types Definitions

rgomezp committed:
Types File Path in Package.json

rgomezp committed:
Nit: rename argument

emawby committed:
Merge pull request #1073 from OneSignal/fix/remove_valid_archs

Remove VALID_ARCHS as User-Defined setting on RCTOneSignal

rgomezp committed:
Fix Notification Id Bug where Notification Obj is Undefined

rgomezp committed:
Make Complete Argument Optional in TS Typings

jkasten2 committed:
Removed Android OneSignal manifest placeholders

* The `onesignal_app_id` and `onesignal_google_project_number` placehold
values were droped from OneSignal-Android-SDK 4.0.0

jkasten2 committed:
Merge pull request #1105 from OneSignal/clean/android_remove_droped_onesignal_manifest_placehold_values

Cleanup - Android - Remove dropped onesignal manifest placeholder values

rgomezp committed:
Rename logLevel method from `onesignal_Log` to `onesignalLog`

Motivation: the native iOS SDK changed the method name `onesignal_Log` to `onesignalLog` for consistency w/the native Android SDK method.

rgomezp committed:
Pass Potential external Id method errors to JS via bridge

Motivation: since we have added user verification to external id setting and modification, we want to surface the error via the function callback by passing it to the JS side in case the error has relevant details.

E.g: "External User Id authentication (auth token) is set to REQUIRED for this application. Please provide an auth token from your backend server or change the setting in the OneSignal dashboard."

rgomezp committed:
Use Major Release Native SDK Versions

Upgrade to use version 4 of the Android SDK, version 3 of the iOS SDK

rgomezp committed:
Update outcome event symbol to OSOutcomeEvent

Native SDK renamed the outcome event variable to `OSOutcomeEvent`

rgomezp committed:
Rename `cancelNotification` to `removeNotification`

Motivation: this function was renamed to be clearer

rgomezp committed:
Update Version Number to Beta 5

rgomezp committed:
Remove TO DO comment

rgomezp committed:
Update Example App Dependencies

rgomezp committed:
Remove Arrow Functions in Favor of Normal Function

Motivation: older devices may have older versions of Chrome which don't have support for ES6 syntax

rgomezp committed:
Small Nit: We Should Allow for the setting of blank strings as values

Motivation: tags can be removed by sending a blank string. Here we're updating the logic to allow for that

rgomezp committed:
Make `postNotification` callbacks optional

rgomezp committed:
Remove Requirement of a Callback to the `logoutEmail` function

Motivation: this should never have been required

Jeasmine committed:
Merge branch 'master' into feature/merge-master

Jeasmine committed:
Add external Id overloading method to index.js

Jeasmine committed:
Merge pull request #1110 from OneSignal/feature/merge-master

Feature/merge master

rgomezp committed:
Update Typescript Types

Motivation:
Rename `cancelNotification` to `removeNotification` for clarity

rgomezp committed:
Add authHash String as optional param in external user id function sig

Motivation: update the types to reflect recent changes for user verification for setExternalUserId

rgomezp committed:
Fix iOS permission status comment in types

Motivation: In the typings, the values for Authorized and Denied were backwards

eraykose committed:
Platform.OS assign fix

Platform.OS is assigned to 'ios' in OSNotification.js and it affects all android settings in apps.

rgomezp committed:
Nil checks in iOS bridge for outcomes methods

Motivation: there are multiple places in native `OneSignalOutcomeEventsController` where we call `success(nil)`. One example is calling `sendUniqueOutcome` multiple times. We should check that the `outcome` object is non-nil before passing it back to JS with the `jsonRepresentation` native method.

rgomezp committed:
Add native logging to outcomes methods in iOS bridge

Motivation: 
- [Context](https://bit.ly/2WJbxHG)
- Adds some logging in the iOS bridge to address the case that `outcome` object is nil

rgomezp committed:
Add logging in Android bridge for cases where outcome object is null

rgomezp committed:
Refactor to match iOS

Motivation: we want Android bridge to match the iOS bridge. For this reason, we are removing the invocation of the callback altogether in the case that the outcome object is null.

Commit:
Typescript Example App (#1128)

* Initial Commit - Entire TS Example App

* Update Yarn Lock to use latest Beta (beta 5)

* Delete Non-TS Example App

Motivation: moving forward we will only maintain the TS version

* Refactor TS Example App

Includes buttons for many functions

* OSConsole to log events and variables

* Finished Setting up Demo App

* Update .gitignore

Motivation: update .gitignore to ignore files from new example app

* Commit XCode Config & Files

* Commit Info.plist, Podfile lock file, and other project files

* Add Error Handling for Getting Trigger Value For Key

Motivation: Example app - add error handling for rejected `getTriggerValueForKey` promise.

Also, remove JSON.stringify around second param to logging function as value is already stringified automatically inside that function.

* Add Clear Handlers to Component Will Unmount

Motivation: to avoid aggregating handlers/listeners, it is good practice to remove them before a component unmounts in the `componentWillUnmount`

* Cross-Platform Conditional Formatting for Console Text

* Add Location Permissions to Config Files

Motivation: we want the example app to work with location out of the box. See reference: https://documentation.onesignal.com/docs/sdk-reference#location-data

Commit:
4.0.0 Release (#1131)

* Update package.json files to 4.0.0

* Update Native SDK Versions

* Remove native iOS files

Motivation: these files were previously there in order to support non-cocoapods integrations. We are removing it here in the major release and will no longer be maintained moving forward.

The manual linking process is likely done by a small portion of developers and introduces more room for error. Cocoapods is pretty standard at this point and can always be added back later or via SwiftPM in the future.
